src/column.rs

```diff
@@ -0,0 +1,105 @@
+use core::marker::PhantomData;
+use core::ptr::NonNull;
+
+use alloc::vec::Vec;
+
+use crate::entities::EntityMeta;
+use crate::{Archetype, Component, ComponentError, Entity, MissingComponent};
+
+/// Column of a single component
+pub struct Column<'a, T> {
+    entities: &'a [EntityMeta],
+    archetypes: &'a [Archetype],
+    archetype_column_offsets: Vec<Option<usize>>,
+    _marker: PhantomData<T>,
+}
+
+impl<'a, T: Component> Column<'a, T> {
+    pub(crate) fn new(
+        entities: &'a [EntityMeta],
+        archetypes: &'a [Archetype],
+        archetype_column_offsets: Vec<Option<usize>>,
+        _marker: PhantomData<T>,
+    ) -> Self {
+        Self {
+            entities,
+            archetypes,
+            archetype_column_offsets,
+            _marker,
+        }
+    }
+
+    /// Get this column's component from entity if it exist
+    pub fn get(&self, entity: Entity) -> Result<&T, ComponentError> {
+        unsafe {
+            let meta = self
+                .entities
+                .get(entity.id as usize)
+                .expect("no such entity");
+            let archetype = self
+                .archetypes
+                .get(meta.location.archetype as usize)
+                .unwrap();
+            let state = self.archetype_column_offsets[meta.location.archetype as usize]
+                .ok_or_else(MissingComponent::new::<T>)?;
+            let target = NonNull::new_unchecked(
+                archetype
+                    .get_base::<T>(state)
+                    .as_ptr()
+                    .add(meta.location.index as usize),
+            );
```

Comment:
`NonNull` is only useful when we're storing it somewhere that might allow a niche to be exposed. Just constructing and immediately destructuring it with `as_ptr` doesn't accomplish anything.